### PR TITLE
fix #1318, fix MinClock not updated fast enough for slow stream

### LINKED ISSUE #1318: MinClock is not updated fast enough for slow stream
For slow stream, like 1 message in 1 second, the MinClock is not udpated fast enough.

### PR DESCRIPTION
**Pull request by manuzhang**
[Review on Reviewable](https://reviewable.io/reviews/gearpump/gearpump/1705)

### CODE REVIEW

**Comment by clockfly** [comment]
Hi @manuzhang 

Can you try to write a app like this to verify whether it works:
1. 10 layer application.
A ->B ->C ->D ->E ->F ->G ->H ->I ->J
2. 1000 tasks for each layer.
3. Previous layer randomly shuffle data to next layer.
4. For the entry layer, each task will generate 1 message in 1 second. 

Expectation:
1. clock can advance fast.

Other concerns:
1. Perfomance.
2. Whether it will create a flooding scenario and cause OOM, as the upstream are sending unsolicited messages.

**Comment by clockfly** [comment]
Please attach the result when you have done the test.

**Comment by manuzhang** [comment]
@clockfly 

I did a test on SOL with the arguments `-streamProducer 1000 -streamProcessor 1000 -stages 10` and added a log to get the update of min clock. Here are the logs.

## With this PR

```
[INFO] [12/17/2015 09:49:58.148] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:00.433] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:02.755] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:04.913] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:07.089] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:09.300] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:11.395] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:13.603] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:15.710] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:17.842] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:19.965] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:22.234] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:24.607] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:26.811] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:29.043] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:31.269] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:33.465] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:35.856] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:38.034] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:40.292] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:42.651] [ClockService] min clock 0
[WARN] [12/17/2015 09:50:44.329] [HealthChecker] Clock has not advanced for 60 seconds since (system clock: Thu Dec 17 09:49:44 CST 2015, app clock: 0)...
[INFO] [12/17/2015 09:50:44.414] [HealthChecker] Stalling Tasks: TaskId(4,42),TaskId(4,188)
[INFO] [12/17/2015 09:50:45.008] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:47.256] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:49.473] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:52.066] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:54.340] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:56.429] [ClockService] min clock 0
[INFO] [12/17/2015 09:50:58.641] [ClockService] min clock 0
[INFO] [12/17/2015 09:51:00.886] [ClockService] min clock 0
[INFO] [12/17/2015 09:51:03.110] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:05.438] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:07.524] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:09.818] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:11.938] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:14.022] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:16.100] [ClockService] min clock 1450316997493
[INFO] [12/17/2015 09:51:18.208] [ClockService] min clock 1450316997495
[INFO] [12/17/2015 09:51:20.325] [ClockService] min clock 1450316997495
[INFO] [12/17/2015 09:51:22.423] [ClockService] min clock 1450316997495
[INFO] [12/17/2015 09:51:24.522] [ClockService] min clock 1450316997506
[INFO] [12/17/2015 09:51:26.590] [ClockService] min clock 1450317039551
[INFO] [12/17/2015 09:51:28.657] [ClockService] min clock 1450317059821
[INFO] [12/17/2015 09:51:30.759] [ClockService] min clock 1450317066812
[INFO] [12/17/2015 09:51:32.830] [ClockService] min clock 1450317069963
[INFO] [12/17/2015 09:51:34.892] [ClockService] min clock 1450317072184
[INFO] [12/17/2015 09:51:36.970] [ClockService] min clock 1450317074284
[INFO] [12/17/2015 09:51:39.042] [ClockService] min clock 1450317076288
[INFO] [12/17/2015 09:51:41.171] [ClockService] min clock 1450317078443
[INFO] [12/17/2015 09:51:43.321] [ClockService] min clock 1450317080672
[INFO] [12/17/2015 09:51:45.419] [ClockService] min clock 1450317082988
[INFO] [12/17/2015 09:51:47.517] [ClockService] min clock 1450317084991
[INFO] [12/17/2015 09:51:49.581] [ClockService] min clock 1450317086995
[INFO] [12/17/2015 09:51:51.684] [ClockService] min clock 1450317089025
[INFO] [12/17/2015 09:51:53.767] [ClockService] min clock 1450317091258
[INFO] [12/17/2015 09:51:55.903] [ClockService] min clock 1450317093371
[INFO] [12/17/2015 09:51:57.977] [ClockService] min clock 1450317095376
[INFO] [12/17/2015 09:52:00.049] [ClockService] min clock 1450317097505
[INFO] [12/17/2015 09:52:02.113] [ClockService] min clock 1450317099427
[INFO] [12/17/2015 09:52:04.201] [ClockService] min clock 1450317101953
[INFO] [12/17/2015 09:52:06.325] [ClockService] min clock 1450317103957
[INFO] [12/17/2015 09:52:08.395] [ClockService] min clock 1450317106036
[INFO] [12/17/2015 09:52:10.479] [ClockService] min clock 1450317108040
[INFO] [12/17/2015 09:52:12.569] [ClockService] min clock 1450317110071
[INFO] [12/17/2015 09:52:14.703] [ClockService] min clock 1450317112118
[INFO] [12/17/2015 09:52:16.811] [ClockService] min clock 1450317114122
[INFO] [12/17/2015 09:52:18.879] [ClockService] min clock 1450317116326
[INFO] [12/17/2015 09:52:20.960] [ClockService] min clock 1450317118330
[INFO] [12/17/2015 09:52:23.018] [ClockService] min clock 1450317120447
[INFO] [12/17/2015 09:52:25.088] [ClockService] min clock 1450317122494
[INFO] [12/17/2015 09:52:27.193] [ClockService] min clock 1450317124680
[INFO] [12/17/2015 09:52:29.308] [ClockService] min clock 1450317126865
[INFO] [12/17/2015 09:52:31.373] [ClockService] min clock 1450317128886
[INFO] [12/17/2015 09:52:33.533] [ClockService] min clock 1450317131119
[INFO] [12/17/2015 09:52:35.618] [ClockService] min clock 1450317133169
[INFO] [12/17/2015 09:52:37.715] [ClockService] min clock 1450317135186
[INFO] [12/17/2015 09:52:39.790] [ClockService] min clock 1450317137371
[INFO] [12/17/2015 09:52:41.865] [ClockService] min clock 1450317139498
[INFO] [12/17/2015 09:52:43.962] [ClockService] min clock 1450317141502
[INFO] [12/17/2015 09:52:46.070] [ClockService] min clock 1450317143506
[INFO] [12/17/2015 09:52:48.173] [ClockService] min clock 1450317145510
[INFO] [12/17/2015 09:52:50.235] [ClockService] min clock 1450317147629
[INFO] [12/17/2015 09:52:52.311] [ClockService] min clock 1450317149743
[INFO] [12/17/2015 09:52:54.406] [ClockService] min clock 1450317151938
[INFO] [12/17/2015 09:52:56.471] [ClockService] min clock 1450317153942
[INFO] [12/17/2015 09:52:58.547] [ClockService] min clock 1450317155947
[INFO] [12/17/2015 09:53:00.676] [ClockService] min clock 1450317158176
[INFO] [12/17/2015 09:53:02.785] [ClockService] min clock 1450317160184
[INFO] [12/17/2015 09:53:04.871] [ClockService] min clock 1450317162212
[INFO] [12/17/2015 09:53:06.975] [ClockService] min clock 1450317164441
[INFO] [12/17/2015 09:53:09.045] [ClockService] min clock 1450317166450
[INFO] [12/17/2015 09:53:11.143] [ClockService] min clock 1450317168586
```

### Without this PR
```
[INFO] [12/17/2015 10:00:53.930] [ClockService] min clock 0
[INFO] [12/17/2015 10:00:56.131] [ClockService] min clock 0
[INFO] [12/17/2015 10:00:58.294] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:00.407] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:02.653] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:04.859] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:06.993] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:09.083] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:11.238] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:13.361] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:15.554] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:17.841] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:20.090] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:22.232] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:24.422] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:26.589] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:28.744] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:30.840] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:32.982] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:35.133] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:37.281] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:39.438] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:41.700] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:43.931] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:46.145] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:48.241] [ClockService] min clock 0
[INFO] [12/17/2015 10:01:50.303] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:01:52.401] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:01:54.515] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:01:56.578] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:01:58.687] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:00.810] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:02.877] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:04.953] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:07.022] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:09.080] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:11.141] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:13.223] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:15.286] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:17.393] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:19.591] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:21.671] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:23.736] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:25.900] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:28.049] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:30.177] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:32.286] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:34.371] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:36.453] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:38.577] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:40.643] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:42.757] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:44.837] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:46.920] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:49.002] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:51.093] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:53.186] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:55.263] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:57.350] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:02:59.442] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:01.519] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:03.577] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:05.645] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:07.711] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:09.812] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:11.935] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:14.093] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:16.181] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:18.274] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:20.355] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:22.416] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:24.517] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:26.603] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:28.713] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:30.795] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:32.868] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:34.971] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:37.041] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:39.111] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:41.225] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:43.301] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:45.393] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:47.518] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:49.670] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:51.807] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:53.906] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:55.985] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:03:58.056] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:00.116] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:02.249] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:04.320] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:06.377] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:08.470] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:10.565] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:12.661] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:14.830] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:16.927] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:19.037] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:21.160] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:23.292] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:25.375] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:27.452] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:29.525] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:31.592] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:33.696] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:35.794] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:37.877] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:39.962] [ClockService] min clock 1450317653865
[WARN] [12/17/2015 10:04:41.311] [HealthChecker] Clock has not advanced for 120 seconds since (system clock: Thu Dec 17 10:02:41 CST 2015, app clock: 1450317653865)...
[INFO] [12/17/2015 10:04:41.319] [HealthChecker] Stalling Tasks: TaskId(0,995)
[INFO] [12/17/2015 10:04:42.070] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:44.139] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:46.209] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:48.314] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:50.406] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:52.496] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:54.627] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:56.688] [ClockService] min clock 1450317653865
[INFO] [12/17/2015 10:04:58.749] [ClockService] min clock 1450317653865
```

You can see that after a slow start for both, the min clock advances fast with the PR but stalls without.

**Comment by clockfly** [comment]
Can you try to reduce the source message frequency to 1 msg/second, and see what happens?

**Comment by manuzhang** [comment]
I forgot to mention that I tested with I msg/second

**Comment by clockfly** [comment]
Can you visulize it with a chart?

**Comment by clockfly** [comment]
Can you justify whether it will create a flooding problem, and how this patch works for this case?

**Comment by manuzhang** [comment]
Here is the visual chart. I aligned them to the start of their application clocks. 

screenshot from 2015-12-17 10 56 27

**Comment by manuzhang** [comment]
@clockfly  updated. AckRequest is sent with the following constraints. 

1. min clock at current task is not updated
2. only send AckRequest to subscription which allows sending more messages (subscription is healthy)
3. only send AckRequest to target tasks whose min clock equals min clock at current task (tasks who are slow)
4. only send AckRequest when pending message count is between 0 and ONE_ACKREQUEST_EVERY_MESSAGE_COUNT / 2 (really slow stream)

With the above, I think we can avoid flooding problem.

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 1]
```diff
@@ -242,6 +243,16 @@ class TaskActor(
       receiveMessage(inputMessage, sender)
     case upstream@ UpstreamMinClock(upstreamClock) =>
       this.upstreamMinClock = upstreamClock
+      val curMin = minClockAtCurrentTask
+      if (curMin == curTaskMinClock) {
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: It is confusing to see both minClockAtCurrentTask and curTaskMinClock.

what is the difference?

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 2]
```diff
@@ -242,6 +243,16 @@ class TaskActor(
...
+        subscriptions.foreach { case (_, subscription) =>
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: This is the hot path, maybe using while is better(foreach is not natively support by java)

**Comment by clockfly** [comment]
```
With the above, I think we can avoid flooding problem.
```
I can see you are optimistic:) But I need it to be right in Math, not only in experience. 
For the input rate m msg/second, there should be a upper bound of rate for the all control messages(ackRequest, ack....)

**Comment by manuzhang** [comment]
@clockfly updated.

**Comment by clockfly** [comment]
Have you estimated the worst case performance?

Suppose the message count is N, how many AckRequest message will be generated for the worst case?

**Comment by manuzhang** [comment]
The number of AckRequest for the worst case would be `O(n)` which is one AckRequest for one message but some could be lost.

**Comment by clockfly** [comment]
I didn't get the same number as you.

If the downstream has not yet replied, so the upstream will send over and over? Possibly generate inifnite AckRequest for one downstream task?

**Comment by manuzhang** [comment]
@clockfly  another try.

**Comment by clockfly** [comment]
Sorry, I have ont found you changed anything

**Comment by clockfly** [comment]
Since this change risky, I suggest to push it to next iteration. 

For now, I don't have confidence to mege it.

**Comment by clockfly** [comment]
several major concerns:
1.  private val allowForMoreAckRequest: Array[Boolean] = Array.fill(parallelism)(false) will take significant memorys. As the dimension is big.  It reprensents dimension * dimension elements.
We shall not introduce additionaly array unless absolutely necessary.

2. How is the wost case performance.

3 what is the computation overhead.

4. whether it is logically right
```
allowForMoreAckRequest(i) &&
 +          pendingMessageCount(i) > 0 &&
 +          pendingMessageCount(i) < maxPendingMessageCount
```
This check doesn't looks reasonble to me.

**Comment by manuzhang** [comment]
ok, let's defer this till we will find a better solution.

**Comment by manuzhang** [comment]
@clockfly I'm back !!!

I extend `AckRequest` with a `countIn: Boolean` field. We check for stalling downstream tasks when receiving `upstreamMinClock`, and send an `AckRequest` with `countIn = true` to the stalling task **if allowed**. This special `AckRequest` is **counted in** like a normal message at both sender (`messageCount`, `pendingMessageCount`) and receiver (`receiveMsgCount`) so its throughput is  bounded by `MAX_PENDING_MESSAGE_COUNT`. All other `AckRequest`s with `countIn = false` are not counted. 

I've verified against your settings


> Can you try to write a app like this to verify whether it works:
1. 10 layer application.
A ->B ->C ->D ->E ->F ->G ->H ->I ->J
2. 1000 tasks for each layer.
3. Previous layer randomly shuffle data to next layer.
4. For the entry layer, each task will generate 1 message in 1 second.

> Expectation:
1. clock can advance fast.

**Comment by clockfly** [comment]
@manuzhang 

Good try!

Instead of defining two AckRequest (some increase message count, some don't), how about we always tag AckRequest message with a message count (same as ONE_ACKREQUEST_EVERY_MESSAGE_COUNT default value: 100) 

Then it has several advantages:
1. We can also detect AckRequest message loss.
2. We don't need to change the AckRequest format.
3. The worst case AckRequest count per message is bounded by "MAX_PENDING_MESSAGE_COUNT / ONE_ACKREQUEST_EVERY_MESSAGE_COUNT"

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 3]
```diff
@@ -285,10 +303,14 @@ class TaskActor(
       doHandleMessage()
   }
 
-  def minClock: TimeStamp = {
-    Math.max(life.birth, Math.min(upstreamMinClock, minClockAtCurrentTask))
-  }
+  /**
+   * @return min clock of this task
+   */
+  def getMinClock: TimeStamp = thisMinClock
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Why not use def minClock: TimeStamp = _minClock

It is confusing to see thisMinClock

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala [hunk 4]
```diff
@@ -189,6 +185,30 @@ class Subscription(
...
+        updateMessageCount(i)
+        sendAckRequest(i, countIn = true)
+      }
+    }
+  }
+
+  private def sendAckRequest(partition: Int, countIn: Boolean = false): Unit = {
+    val targetTask = TaskId(processorId, partition)
+    val ackRequest = AckRequest(taskId, messageCount(partition), sessionId, countIn)
+    LOG.debug(s"sending ack request $ackRequest to $targetTask")
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: This is on a critical path. 
LOG.debug can hurt your performance in unexpected way.

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 5]
```diff
@@ -242,8 +237,31 @@ class TaskActor(
       receiveMessage(inputMessage, sender)
     case upstream@ UpstreamMinClock(upstreamClock) =>
       this.upstreamMinClock = upstreamClock
-      val latestMinClock = minClock
-      val update = UpdateClock(taskId, latestMinClock)
+      val minClocks = subscriptions.map { case (id, sub) => sub -> sub.minClock }.toMap
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: This change is over-complicated considering what it want to do.

It use **4** rounds! (map, foldLeft, Math.min, foreach)

Really necessary? Please try to use one round to fix this.

**Comment by manuzhang** [comment]
@clockfly  updated against your comments

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 6]
```diff
@@ -344,6 +354,7 @@ object TaskActor {
     def generateAckResponse(ackRequest: AckRequest, sender: ActorRef): Ack = {
       val sessionId = ackRequest.sessionId
       if (receivedMsgCount.containsKey(sessionId)) {
+        receivedMsgCount.put(sessionId, (receivedMsgCount.get(sessionId) + 1).toShort)
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: I mean + ONE_ACKREQUEST_EVERY_MESSAGE_COUNT

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 7]
```diff
@@ -242,19 +237,30 @@ class TaskActor(
...
     case upstream@ UpstreamMinClock(upstreamClock) =>
       this.upstreamMinClock = upstreamClock
-      val latestMinClock = minClock
-      val update = UpdateClock(taskId, latestMinClock)
+
+      val subMinClock = subscriptions.foldLeft(Long.MaxValue) {
+        case (min, (_, sub)) =>
+          val subMin = sub.minClock
+          if (subMin == _minClock) {
+            sub.sendAckRequestToStallingTask(_minClock)
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Please add one line doc here for "what subMin == _minClock means"

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala [hunk 8]
```diff
@@ -140,8 +137,7 @@ class Subscription(
   private def flush: Unit = {
     lastFlushTime = System.currentTimeMillis()
     allTasks.foreach { targetTaskId =>
-      val ackRequest = AckRequest(taskId, messageCount(targetTaskId.index), sessionId)
-      transport.transport(ackRequest, targetTaskId)
+      sendAckRequest(targetTaskId.index)
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: Why not updating the message count here???

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala [hunk 9]
```diff
@@ -242,19 +237,30 @@ class TaskActor(
       receiveMessage(inputMessage, sender)
     case upstream@ UpstreamMinClock(upstreamClock) =>
       this.upstreamMinClock = upstreamClock
-      val latestMinClock = minClock
-      val update = UpdateClock(taskId, latestMinClock)
+
+      val subMinClock = subscriptions.foldLeft(Long.MaxValue) {
+        case (min, (_, sub)) =>
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Can we avoid this unapply call?

**Comment by clockfly** [comment]
Please also do so some performance regression test

File: streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala (hunk 6 above)
Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: `receivedMsgCount.get(sessionId) + ONE_ACKREQUEST_EVERY_MESSAGE_COUNT` when receiving `AckRequest`?

Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Yes.

Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Or name another constant which have same value as ONE_ACKREQUEST_EVERY_MESSAGE_COUNT

Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: why ? This breaks message check unless we do the same at sender's side.

Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: Yes, we should do the same at the sender side.

The purpose is to reduce the AckRequest count.

Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/TaskActor.scala]: ok, I see your point.

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala [hunk 10]
```diff
@@ -189,13 +185,38 @@ class Subscription(
...
     maxPendingCount = Shorts.max(pendingMessageCount: _*)
   }
+
 }
 
 object Subscription {
   //make sure it is smaller than MAX_PENDING_MESSAGE_COUNT
   final val ONE_ACKREQUEST_EVERY_MESSAGE_COUNT = 100
   final val MAX_PENDING_MESSAGE_COUNT = 1000
+  final val ACKREQUEST_INCREMENT_COUNT = 100
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: NOTE, this is default value. The actual value used is at: 
```
 class Subscription(
     appId: Int,		     appId: Int,
     executorId: Int,		     executorId: Int,
     taskId: TaskId,		     taskId: TaskId,
     subscriber: Subscriber, sessionId: Int,		     subscriber: Subscriber, sessionId: Int,
     transport: ExpressTransport,		     transport: ExpressTransport,
     maxPendingMessageCount: Int = MAX_PENDING_MESSAGE_COUNT,		     maxPendingMessageCount: Int = MAX_PENDING_MESSAGE_COUNT,
     ackOnceEveryMessageCount: Int = ONE_ACKREQUEST_EVERY_MESSAGE_COUNT) {		     ackOnceEveryMessageCount: Int = ONE_ACKREQUEST_EVERY_MESSAGE_COUNT) {
```

You need to make sure "ACKREQUEST_INCREMENT_COUNT" always match the effective ackOnceEveryMessageCount.

So may be we should just use "ackOnceEveryMessageCount".

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala [hunk 11]
```diff
@@ -189,13 +185,38 @@ class Subscription(
...
+  private def sendAckRequest(partition: Int): Unit = {
+    // we increment more count for each AckRequest
+    // to throttle the number of unacked AckRequest
+    incrementMessageCount(partition, ACKREQUEST_INCREMENT_COUNT)
+    val targetTask = TaskId(processorId, partition)
+    val ackRequest = AckRequest(taskId, messageCount(partition), sessionId)
+    transport.transport(ackRequest, targetTask)
+  }
+
+  private def incrementMessageCount(partition: Int, count: Int = 1): Unit = {
```
Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: Maybe not necessary to use default parameter value.

File: streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala (hunk 10 above)
Comment (manuzhang) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: but we can not use `ackOnceEveryMessageCount` at downstream tasks.

Comment (clockfly) [inline streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala]: They share the same application config.

**Comment by clockfly** [comment]
+1

**Pull request merged**